domains/client/domain-executor/src/domain_bundle_processor.rs

```diff
@@ -384,6 +525,142 @@ where
+                // Use the origin fork_choice for the target primary block,
+                // the intermediate ones use `Custom(false)`.
+                let fork_choice = if i == primary_imports.len() - 1 {
+                    fork_choice
+                } else {
+                    ForkChoiceStrategy::Custom(false)
+                };
+
+                domain_parent = self
+                    .process_bundles_at(
+                        (primary_info.hash, primary_info.number, fork_choice),
+                        domain_parent,
+                    )
+                    .await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn process_bundles_at(
+        &self,
+        primary_info: (PBlock::Hash, NumberFor<PBlock>, ForkChoiceStrategy),
+        parent_info: (Block::Hash, NumberFor<Block>),
+    ) -> Result<(Block::Hash, NumberFor<Block>), sp_blockchain::Error> {
+        let (primary_hash, primary_number, fork_choice) = primary_info;
+        let (parent_hash, parent_number) = parent_info;
+
+        tracing::debug!(
+                "Building a new domain block derived from primary block #{primary_number},{primary_hash} \
+                on top of #{parent_number},{parent_hash}"
+            );
+
+        let (bundles, shuffling_seed, maybe_new_runtime) =
+            preprocess_primary_block(self.domain_id, &*self.primary_chain_client, primary_hash)?;
+
+        let extrinsics = self.bundles_to_extrinsics(parent_hash, bundles, shuffling_seed)?;
+
+        let digests = {
+            let at = if self.domain_id.is_system() {
+                translate_block_hash_type::<Block, SBlock>(parent_hash)
+            } else {
+                // include the latest state root of the system domain
+                self.system_domain_client.info().best_hash
+            };
+            let mut digest = Digest::default();
+            if let Some(state_root_update) = self.system_domain_state_root_update_digest(at)? {
+                digest.push(state_root_update);
+            }
+            if self.domain_id.is_system() {
+                let primary_block_info =
+                    DigestItem::primary_block_info((primary_number, primary_hash));
+                digest.push(primary_block_info);
+            }
+            digest
+        };
+
+        let domain_block_result = self
+            .process_domain_block(
+                (primary_hash, primary_number),
+                (parent_hash, parent_number),
+                extrinsics,
+                maybe_new_runtime,
+                fork_choice,
+                digests,
+            )
+            .await?;
+
+        let at = if self.domain_id.is_system() {
+            translate_block_hash_type::<PBlock, ParentChainBlock>(primary_hash)
+        } else {
+            // TODO: just make it compile for now, likely wrong, rethink about it.
+            self.parent_chain.info().best_hash
+        };
```

Comment:
IMHO, here we need to get the hash of the parent chain block that drives from the primary block, once we get that hash both system/core domains can use it (though the system domain has a fast path that uses the `primary_hash `  directly).